src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,491 @@
+                await writer.WriteBinaryStringAsync(value, buffer, arrayPool).ConfigureAwait(false);
+                await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">Byte array to be written.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteBinaryStringAsync(this TextWriter writer, byte[] value, Ref<char[]> buffer, ICharArrayPool arrayPool)
+        {
+            Debug.Assert(writer != null, "writer != null");
+            Debug.Assert(value != null, "The value must not be null.");
+
+            buffer.Value = BufferUtils.InitializeBufferIfRequired(arrayPool, buffer.Value);
+            Debug.Assert(buffer.Value != null);
+
+            int bufferLength = buffer.Value.Length;
+
+            // Try to hold base64 string as much as possible in one converting.
+            int bufferByteSize = bufferLength * 3 / 4;
+
+            for (int offsetIn = 0; offsetIn < value.Length; offsetIn += bufferByteSize)
+            {
+                int length = bufferByteSize;
+                if (offsetIn + length > value.Length)
+                {
+                    length = value.Length - offsetIn;
+                }
+
+                int output = Convert.ToBase64CharArray(value, offsetIn, length, buffer.Value, 0);
+                await writer.WriteAsync(buffer.Value, 0, output).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Returns the string value with special characters escaped.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="inputString">Input string value.</param>
+        /// <param name="stringEscapeOption">The string escape option.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="bufferPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteEscapedJsonStringAsync(this TextWriter writer, string inputString,
+            ODataStringEscapeOption stringEscapeOption, Ref<char[]> buffer, ICharArrayPool bufferPool = null)
+        {
+            Debug.Assert(writer != null, "writer != null");
+            Debug.Assert(inputString != null, "The string value must not be null.");
+
+            await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+            await writer.WriteEscapedJsonStringValueAsync(inputString, stringEscapeOption, buffer, bufferPool).ConfigureAwait(false);
+            await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);

```

Comment:
Since you've created `WriteQuotedAsync` to reduce the number of async calls when writing something wrapped into quotes, can that approach be applied here as well? If it's worth while performance wise.
Don't know whether the quotes could be concatenated around the inputString, or whether the quotes could be added to the buffer that gets written.

Reply:
@habbes Based on the fact that quotes in the input string would be escaped, concatenating the enclosing quotes would result into them getting escaped as well. Regarding the option of writing the quotes to directly to the buffer, that's a change I'd prefer to be handle separately as part of perf improvements back by an actual performance analysis.

Reply:
If "WriteEscapedJsonStringValueAsync" is only called from here (or from other places where the string must be enclosed in quotes), then we could add the writing of the quote characters as part of that method, rather than writing them to the buffer before & after calling that method.  But, happy to take that separately, as it looks like the WriteEscapedJsonStringValueAsync method could use stringbuilder to build the string and writeasync once, but we'd want to do perf to verify that using a stringbuilder would be better than individual async writes (though I suspect it would be).

---
In reply to: [584965109](https://github.com/OData/odata.net/pull/2002#discussion_r584965109) [](ancestors = 584965109)